feat: OPTIC-1938: Proxy for storages when presigned urls are off

makseq opened the pull request:
# Storage proxy added

**Our current implementation with presigned urls:**

![image](https://github.com/user-attachments/assets/a00af1f2-9b9f-4a1a-ab63-a590c41f547a)

**This PR adds proxying from storages instead of redirecting to presigned urls:**

![image](https://github.com/user-attachments/assets/55065d6c-4077-4e01-8e37-4cd910102c0d)


# Minor fixes

Fixed race condition on Project => Settings => Cloud Storages page when project=undefined on the first page loading. 

![image](https://github.com/user-attachments/assets/1b242b9c-80c6-45b6-8f1a-eef7a31b9bc2)

makseq committed:
feat: OPTIC-1938: Proxy for storages when presigned urls are off

makseq committed:
Add first version of good implementation

makseq committed:
Revert

makseq committed:
Fixes

makseq committed:
Fix

makseq committed:
Fix

makseq committed:
Always generate /presign

makseq committed:
Redesign

makseq committed:
Fix

makseq committed:
Move resover s3 to storage mixin

farioas commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14364600032)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14364600032

makseq committed:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Merge branch 'fb-optic-1938' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Fix storage list project undefined

makseq committed:
Remove useEffect

makseq committed:
Move gcs to storagemixin

makseq committed:
Fix azure

makseq committed:
Fixes

makseq committed:
Fixes

makseq committed:
Remove fflag_fix_all_lsdv_4711_cors_errors_accessing_task_data_short

makseq committed:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Merging

makseq committed:
Build

makseq committed:
Fix

makseq committed:
test_presign_storage_data.py fixed

makseq committed:
Turn on ff

makseq committed:
Fix test with 1

makseq committed:
Linters

makseq committed:
Blue

AndrejOros commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14396934665)

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-optic-1938'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14396934665

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14396934665

makseq committed:
Fix test

makseq committed:
Merge branch 'fb-optic-1938' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Fix

bmartel reviewed (approved):
Looks great! The one thing I think that remains is how we align the browsers ability to cache the requests. One part of this whole solution was a special ServiceWorker that caches the resolution of the URI's so that the browser can effectively cache the requests when able, like with images for example. This saves on bandwidth significantly in presigned url use cases with normal usage within DataManager -> QuickView. The ServiceWorker only acted on url's that had the word `presign` in it, but now I see the endpoint is `resolve` instead. This will mean that the previous behaviour will not work completely, so we will have to update the ServiceWorker accordingly as well. I can make that change on this PR.

makseq committed:
Fix

bmartel committed:
serviceworker change to allow the new presign resolving endpoint to be handled the same way as before

bmartel committed:
Merge branch 'fb-optic-1938' of github.com:HumanSignal/label-studio into fb-optic-1938

makseq commented:
@bmartel great point! Thank you for the fix! `/resolve` now makes more sense. I kept `/presign` for backward compatibility.

bmartel committed:
fixed serviceworker cache logic

bmartel committed:
fixed serviceworker cache logic

bmartel committed:
simplify the logic

makseq committed:
Add more unit tests

makseq committed:
Merge branch 'fb-optic-1938' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Linters

makseq committed:
Fix

makseq committed:
Fix

AndrejOros commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14439543164)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14439543164

makseq committed:
Fix cache invalidate

makseq committed:
Merge branch 'fb-optic-1938' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14457159024)
> **Error**: Workflow failed

makseq committed:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Fixes

makseq committed:
Fix blue

makseq committed:
blue

AndrejOros commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14462835950)

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-optic-1938'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14462835950

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14462835950

nikitabelonogov commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14464981467)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14464981467

hakan458 commented on label_studio/io_storages/proxy_api.py:
```diff
@@ -0,0 +1,178 @@
+                return Response(status=status.HTTP_404_NOT_FOUND)
+            presign = storage.presign
+        else:
+            presign = True
```
why do we default to true here?

hakan458 replied:
I guess so that if the FF is not set, we default to the old behavior, correct?

makseq replied:
Yes, you are right, it's default for the old behavior.

hakan458 commented on label_studio/io_storages/proxy_api.py:
```diff
@@ -0,0 +1,178 @@
+            # If we have the data and content type, return the response
+            if data is not None:
+                content_type = content_type or 'application/octet-stream'
+                response = RangedFileResponse(request, data, content_type=content_type)
```
What is RangedFileResponse and why do we need it?

makseq replied:
It's a standard response that supports streaming. We use it here for uploaded files and more - https://github.com/search?q=repo%3AHumanSignal%2Flabel-studio%20RangedFileResponse&type=code

makseq commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14489597608)
> **Error**: Workflow failed

makseq commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14489670052)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14489670052

makseq committed:
Add TimeoutRangedFileResponse

makseq committed:
Linter

makseq committed:
Merge branch 'fb-optic-1938' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14492054959)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14492054959

makseq committed:
Add direct stream from S3

makseq committed:
Merge branch 'fb-optic-1938' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Add normal streaming for s3

makseq committed:
Before iter_chunks

makseq committed:
Working streaming for s3 iter_chunks

makseq committed:
Second working version

makseq committed:
Fixes with working version of streaming s3

jombooth commented on label_studio/io_storages/proxy_api.py:
```diff
@@ -0,0 +1,327 @@
+        
+        try:
+            for chunk in stream_body.iter_chunks(chunk_size=chunk_size):
+                print(f"==> chunk: {chunks_yielded}, {total_bytes}\n")
```
If this is intended for the final version, better to use `logger.info` than `print`

jombooth commented on label_studio/io_storages/proxy_api.py:
```diff
@@ -0,0 +1,327 @@
+            
+            if stream is None:
+                logger.error(f'Failed to get direct stream from storage {storage}')
+                return Response(status=status.HTTP_500_INTERNAL_SERVER_ERROR)
```
I know we do this in a few places, but it seems a bit iffy to me whether we should be intentionally returning 500s from the API; usually these indicate errors that we did not anticipate. How would https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/424 sound to you?

makseq replied:
fixed

jombooth commented on label_studio/io_storages/proxy_api.py:
```diff
@@ -0,0 +1,327 @@
+            time_limited_stream = self.time_limited_chunker(stream)
+            
+            # Set up streaming response with S3's status code
+            status_code = metadata.get('StatusCode', 200)
```
Should it be an error if `StatusCode` wasn't in metadata? Also, is this (as the comment suggests) just an S3 thing, or does it apply to other storage backends?

makseq replied:
Good questions. It should be applicable to all storages. I will fix it

jombooth commented on label_studio/io_storages/proxy_api.py:
```diff
@@ -0,0 +1,327 @@
+        """
+        Proxy the data using iter_chunks directly from storage streaming object.
+        
+        This implementation forwards Range headers to S3 and streams the response
```
same question about S3 here, is this S3 specific?

makseq replied:
fixed. it's for any storages

jombooth commented:
overall LG (and very impressive work), biggest concerns are whether we've introduced S3 specific details to code that's used for other storages as well, as in https://github.com/HumanSignal/label-studio/pull/7354#discussion_r2049528505

makseq committed:
Working cleanup

makseq committed:
Blue & ruff

makseq committed:
Fixes

makseq committed:
Linters

makseq committed:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
fix tests

makseq committed:
Working GCS streaming

makseq committed:
Working Azure

makseq committed:
==> GCS old streaming implementation and new with direct http requests <==

makseq committed:
New GCS streaming over streaming http request

makseq committed:
Fix tests

makseq commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14550330038)

makseq committed:
Tests are fixed

makseq committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14550330038

makseq committed:
Merge branch 'develop' into 'fb-optic-1938'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14550330038

makseq committed:
Add GCS timeout

makseq committed:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Merge branch 'fb-optic-1938' of github.com:heartexlabs/label-studio into fb-optic-1938

makseq committed:
Tests passed. Everything is working

makseq committed:
Revert azure

makseq committed:
Fix for GCS

makseq committed:
Linters

makseq committed:
Fix tests again
